src/shims/env.rs

```diff
@@ -20,15 +21,32 @@ impl EnvVars {
         ecx: &mut InterpCx<'mir, 'tcx, Evaluator<'tcx>>,
         excluded_env_vars: Vec<String>,
     ) {
+        let mut vars = Vec::new();
         if ecx.machine.communicate {
+                // Put each environment variable pointer in `EnvVars`, collect pointers.
             for (name, value) in env::vars() {
                 if !excluded_env_vars.contains(&name) {
                     let var_ptr =
                         alloc_env_var_as_c_str(name.as_ref(), value.as_ref(), ecx);
                     ecx.machine.env_vars.map.insert(OsString::from(name), var_ptr);
+                    vars.push(var_ptr);
                 }
             }
         }
+        // Make an array with all these pointers, in the Miri memory.
+        let tcx = ecx.tcx;
+        let environ_layout =
+            ecx.layout_of(tcx.mk_array(tcx.mk_imm_ptr(tcx.types.u8), vars.len() as u64)).unwrap();
+        let environ_place = ecx.allocate(environ_layout, MiriMemoryKind::Env.into());
+        for (idx, var) in vars.into_iter().enumerate() {
```

Comment:
Isn't there a trailing NULL or so? How does the C code know how many pointers there are?

Reply:
yes there is a trailing null pointer, I'll add it now